project/src/main/ui/level-select/hookable-grade-label.gd

```diff
@@ -0,0 +1,120 @@
+class_name HookableGradeLabel
+extends Node2D
+"""
+Shows a grade label for a specific level.
+
+This can be something like 'S' or 'B+' if a level has been cleared, or something like a lock/key icon for levels
+which haven't yet been cleared.
+"""
+
+# the level button this grade label applies to
+var button: LevelSelectButton setget set_button
+
+var _crown_texture: Texture = preload("res://assets/main/ui/level-select/crown.png")
+var _key_texture: Texture = preload("res://assets/main/ui/level-select/key.png")
+var _locked_texture: Texture = preload("res://assets/main/ui/level-select/locked.png")
+
+"""
+Assigns this label's button, and updates the label's appearance.
+"""
+func set_button(new_button: LevelSelectButton) -> void:
+	button = new_button
+	
+	button.get_node("GradeHook").remote_path = button.get_node("GradeHook").get_path_to(self)
+	if not button.is_connected("lowlight_changed", self, "_on_LevelSelectButton_lowlight_changed"):
+		button.connect("lowlight_changed", self, "_on_LevelSelectButton_lowlight_changed")
+	if not button.is_connected("tree_exited", self, "_on_LevelSelectButton_tree_exited"):
+		button.connect("tree_exited", self, "_on_LevelSelectButton_tree_exited")
+	
+	_refresh_appearance()
+
+
+"""
+Updates the text/icon based on the level's status.
+
+If the level has been cleared, we display text corresponding to the player's grade. If the level hasn't been cleared,
+we show a lock/key/crown icon corresponding to the level's lock status.
+"""
+func _refresh_appearance() -> void:
+	if button.is_class("WorldSelectButton"):
+		var world_button: WorldSelectButton = button
+		var worst_rank := 0.0
+		for rank in world_button.ranks:
+			worst_rank = max(worst_rank, rank)
+		_refresh_label_text(RankCalculator.grade(worst_rank))
+	else:
+		var result := PlayerData.level_history.best_result(button.level_id)
+		if not result:
+			_refresh_status_icon()
+		elif result.compare == "-seconds":
+			_refresh_label_text(RankCalculator.grade(result.seconds_rank))
+		else:
+			_refresh_label_text(RankCalculator.grade(result.score_rank))
+	
+	visible = false if button.lowlight else true
+
+
+"""
+Updates the icon based on the level's status.
+
+We show a lock/key/crown icon corresponding to the level's lock status.
+"""
+func _refresh_status_icon() -> void:
```

Comment:
refresh_status_icon could take a lock status, for consistency with refresh_label_text

Reply:
`refresh_status_icon` now takes a lock status.